app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt

```diff
@@ -46,171 +35,225 @@ class MaskLayer(
+        }
+        currentPath?.let {
+            canvas.drawPath(it, currentPaint)
         }
     }
 
-    private fun drawPathText(
-        path: PointToPointPath,
-        pathPaint: Paint,
-        text: String,
-        canvas: Canvas
-    ) {
-        val textBoundsRuler = Rect()
-        textPaint.getTextBounds(text, 0, text.length, textBoundsRuler)
+    private fun calculateCurrentStrokeWidth() = PATH_STROKE_WIDTH / currentScale
 
-        val currentScale = currentScale()
-        val (firstPointX, firstPointY) = path.firstPoint
-        val textX = firstPointX - textBoundsRuler.width() / 2
-        val verticalPadding = pxToDp(VERTICAL_PADDING_PX * 2) / currentScale
-        val textY = firstPointY +
-                if (path.verticalDirection > 0) {
-                    -(pathPaint.strokeWidth / 2 + verticalPadding)
-                } else {
-                    textBoundsRuler.height() + pathPaint.strokeWidth / 2 +
-                            verticalPadding - pxToDp(VERTICAL_PADDING_PX) / currentScale
-                }
+    private fun refreshPaintBrush() {
+        currentPaintBrush =
+            newDefaultPaintBrush(currentDiseaseStage.maskColor, calculateCurrentStrokeWidth())
+        paintBrushPendingRecreation = false
+    }
 
-        canvas.drawText(text, textX, textY, textStrokePaint)
-        canvas.drawText(text, textX, textY, textPaint)
+    private fun refreshPaintEraser() {
+        currentPaintEraser = newDefaultPaintEraser(calculateCurrentStrokeWidth())
+        paintEraserPendingRecreation = false
     }
 
-    private fun pxToDp(px: Int) = (px * context.resources.displayMetrics.density).toInt()
+    private fun resolveCurrentPaint() {
+        currentPaint = if (isCurrentModeDraw()) currentPaintBrush else currentPaintEraser
+    }
 
-    fun onViewSizeChanged(width: Int, height: Int) {
-        viewDimensions = width to height
-        scaleBrushAndTextPaints()
+    fun drawMode() {
+        currentMode = MaskCustomView.Mode.Draw
     }
 
-    fun onScaleChanged() = scaleBrushAndTextPaints()
+    fun eraseMode() {
+        currentMode = MaskCustomView.Mode.Erase
+    }
 
     fun undo() = undoPendingPaths++
 
     fun redo() = undoPendingPaths--
 
-    fun undoAvailable() = (pathsPaintsAndStagesNames.size - undoPendingPaths) > 0
+    fun undoAvailable() = (pathsAndPaints.size - undoPendingPaths) > 0
 
     fun redoAvailable() = undoPendingPaths > 0
 
     fun getBitmap(): Bitmap {
-        val (width, height) = bitmapDimensions
-        return Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888).apply {
+        return Bitmap.createBitmap(
+            currentStateBitmap.width,
+            currentStateBitmap.height,
+            Bitmap.Config.ARGB_8888
+        ).apply {
             val canvas = Canvas(this)
             drawPaths(canvas, removeAlpha = true)
         }
     }
 
     fun drawStart(x: Float, y: Float) {
+        refreshPaints()
```

Comment:
Shouldn't the `Paint` objects be "refreshed" only when the scale changes or the masking color changes?

Reply:
Yes, but that was a problem in some flows (rotation, for instance). Either way, this method currently only reinstantiates the paints if needed, so in those events you mentioned we just flag them to be recreated, and here we recreate them only once, when they're really going to be used. Makes sense?

Reply:
I was just worried about the potential latency in the drawing action that this might add, but probably it would be a premature optimisation to change it, so I am fine :D